engine/access/ingestion/collection_syncer.go

```diff
@@ -0,0 +1,398 @@
+	// may contain unverifiable guarantees (in case this node has just joined the network)
+	if height <= s.lastFullBlockHeight.Value() {
+		s.logger.Info().
+			Msg("skipping requesting collections for finalized block as its collections have been already retrieved")
+		return
+	}
+
+	s.requestCollections(missingCollections, false)
+}
+
+// requestCollections registers collection download requests in the requester engine,
+// optionally forcing immediate dispatch.
+func (s *CollectionSyncer) requestCollections(missingCollections []*flow.CollectionGuarantee, immediately bool) {
+	for _, collection := range missingCollections {
+		guarantors, err := protocol.FindGuarantors(s.state, collection)
+		if err != nil {
+			// failed to find guarantors for guarantees contained in a finalized block is fatal error
+			s.logger.Fatal().Err(err).Msgf("could not find guarantors for guarantee %v", collection.ID())
+		}
+		s.requester.EntityByID(collection.ID(), filter.HasNodeID[flow.Identity](guarantors...))
+	}
+
+	if immediately {
+		s.requester.Force()
+	}
+}
+
+// updateLastFullBlockHeight updates the next highest block height where all previous collections have been indexed.
+// No errors expected during normal operation.
+func (s *CollectionSyncer) updateLastFullBlockHeight() error {
+	lastFullBlockHeight := s.lastFullBlockHeight.Value()
+	lastFinalizedBlock, err := s.state.Final().Head()
+	if err != nil {
+		return fmt.Errorf("failed to get finalized block: %w", err)
+	}
+
+	// track the latest contiguous full height
+	newLastFullBlockHeight, err := s.findLowestBlockHeightWithMissingCollections(lastFullBlockHeight, lastFinalizedBlock.Height)
+	if err != nil {
+		return fmt.Errorf("failed to find last full block height: %w", err)
+	}
+
+	// if more contiguous blocks are now complete, update db
+	if newLastFullBlockHeight > lastFullBlockHeight {
+		err := s.lastFullBlockHeight.Set(newLastFullBlockHeight)
+		if err != nil {
+			return fmt.Errorf("failed to update last full block height: %w", err)
+		}
+
+		s.collectionExecutedMetric.UpdateLastFullBlockHeight(newLastFullBlockHeight)
+
+		s.logger.Debug().
+			Uint64("last_full_block_height", newLastFullBlockHeight).
+			Msg("updated last full block height counter")
+	}
+
+	return nil
+}
+
+// findLowestBlockHeightWithMissingCollections finds the next block height with missing collections,
+// returning the latest contiguous height where all collections are present.
+//
+// No errors expected during normal operations.
+func (s *CollectionSyncer) findLowestBlockHeightWithMissingCollections(
+	lastKnownFullBlockHeight uint64,
+	finalizedBlockHeight uint64,
+) (uint64, error) {
+	newLastFullBlockHeight := lastKnownFullBlockHeight
+
+	for currBlockHeight := lastKnownFullBlockHeight + 1; currBlockHeight <= finalizedBlockHeight; currBlockHeight++ {
+		missingCollections, err := s.findMissingCollectionsAtHeight(currBlockHeight)
+		if err != nil {
+			return 0, ErrMissingCollection
```

Comment:
similar to `findMissingCollections`, this should just return the error
```suggestion
			return 0, err
```